### PR TITLE
[OnOff] Q implementation for OnTime and OffWaitTime attributes

### PR DESCRIPTION
**Pull request by ReneJosefsen**
#### Summary

This PR reduces the amount of reporting of OnTime and OffWaitTime in the OnOff cluster. The attributes are only reported:

- If the delta change is larger than 10,
- If it reaches 0

The PR also adds an implementation of TC-OO-2.8 which is the proposal for validating the reporting rules.

#### Related issues

Spec: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/12681
Test plan: https://github.com/CHIP-Specifications/chip-test-plans/pull/5849

#### Testing

Passing TC-OO-2.8 against all-clusters running with the changes to the onoff server as proposed in this PR.

Added TestOnTimeAndOffWaitTimeReporting to unit test the implementation

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

### CODE REVIEW

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp [hunk 1]
```diff
@@ -548,16 +550,34 @@ bool OnOffServer::OnWithTimedOffCommand(app::CommandHandler * commandObj, const
     if (currentOffWaitTime > 0 && !isOn)
     {
         uint16_t newOffWaitTime = currentOffWaitTime < offWaitTime ? currentOffWaitTime : offWaitTime;
-        OnOff::Attributes::OffWaitTime::Set(endpoint, newOffWaitTime);
+
+        // If the delta of the change is larger then 10 (1s) the change is reported
+        if (abs(newOffWaitTime - currentOffWaitTime) > 10)
```
Comment (andy31415) [inline src/app/clusters/on-off-server/codegen/on-off-server.cpp]: How about implementations for https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/on-off-server/OnOffLightingCluster.h#L126 ?

Some of these changes are often done with https://github.com/project-chip/connectedhomeip/blob/master/src/app/cluster-building-blocks/QuieterReporting.h

Comment (andy31415) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: For the OnOffLightingCluster changes we also need unit tests added/updated.

File: src/python_testing/TC_OO_2_8.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+#
+
+# See https://github.com/project-chip/connectedhomeip/blob/master/docs/testing/python.md#defining-the-ci-test-arguments
+# for details about the block below.
+#
+# FIXME: https://github.com/project-chip/connectedhomeip/issues/36885
+# === BEGIN CI TEST ARGUMENTS ===
+# test-runner-runs:
+#   run1:
+#     app: ${ALL_CLUSTERS_APP}
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py:25]: all-devices with on-off-light device should probably also be tested

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp [hunk 3]
```diff
@@ -548,16 +550,34 @@ bool OnOffServer::OnWithTimedOffCommand(app::CommandHandler * commandObj, const
...
+        {
+            markOffWaitTimeDirty = app::MarkAttributeDirty::kYes;
```
Comment (andy31415) [inline src/app/clusters/on-off-server/codegen/on-off-server.cpp:553]: We only seem to keep "current time" and compare against that. However what if several commands update this in increments less than 10 but overall eventually the value does change to >10 delta? Should clients not be guaranteed to receive a value when it changes to more than 1s regardless of cause? That probably requires extra storage for the "last value reported".

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp [hunk 4]
```diff
@@ -622,19 +648,23 @@ void OnOffServer::updateOnOffTimeCommand(chip::EndpointId endpoint)
...
             offWaitTime--;
-            OnOff::Attributes::OffWaitTime::Set(endpoint, offWaitTime);
         }
 
         ChipLogDetail(Zcl, "Timer Callback - wait Off Time:  %d", offWaitTime);
 
         // Validate if necessary to restart timer
         if (offWaitTime > 0)
         {
+            // Do not report the regular countdown
```
Comment (andy31415) [inline src/app/clusters/on-off-server/codegen/on-off-server.cpp:654]: should this not be reported every second? 

It seems maybe we should replace the set above with a

```cpp
OnOff::Attributes::OffWaitTime::Set(
    endpoint, 
    offWaitTime,
    (offWaitTime > 0) ? app::MarkAttributeDirty::kNo : app::MarkAttributeDirty::kYes
);
``` 

for smaller PR delta? Or is the current code assumed clearer?

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp [hunk 5]
```diff
@@ -602,16 +622,22 @@ void OnOffServer::updateOnOffTimeCommand(chip::EndpointId endpoint)
...
             onTime--;
-            OnOff::Attributes::OnTime::Set(endpoint, onTime);
         }
 
         if (onTime == 0)
         {
+            // Report that OnTime reached 0
+            OnOff::Attributes::OnTime::Set(endpoint, onTime, app::MarkAttributeDirty::kYes);
+
             ChipLogDetail(Zcl, "Timer callback - Turning off OnOff");
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:628]: This was a miss on my end, I have added the Set back in. Resolving.

File: src/python_testing/TC_OO_2_8.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                         "The second entry in reportedOnTimeValueList is in the range 145 - 150"),
+                TestStep(12, "TH verifies the third entry in reportedOnTimeValueList is 0",
+                         "The third entry in reportedOnTimeValueList is equal to 0"),
+
+                TestStep(13, "TH stores the reported values of OffWaitTime in all incoming reports for OffWaitTime attribute, that contains data in reportedOffWaitTimeValueList and verifies reportedOffWaitTimeValueList contains three entries",
+                         "reportedOffWaitTimeValueList has 3 entries in the list"),
+                TestStep(14, "TH verifies the first entry in reportedOffWaitTimeValueList is approximately 100 (10s)",
+                         "The first entry in reportedOffWaitTimeValueList is in the range 95 - 100"),
+                TestStep(15, "TH verifies the second entry in reportedOffWaitTimeValueList is approximately 150",
+                         "The second entry in reportedOffWaitTimeValueList is in the range 145 - 150"),
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: Addressed

File: src/python_testing/TC_OO_2_8.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+        self.step(11)
+        asserts.assert_less_equal(reportedOnTimeValueList[1].value, 150, "Unexpected second OnTime report")
+        asserts.assert_almost_equal(reportedOnTimeValueList[1].value, 150, delta=10, msg="Unexpected second OnTime report")
+
+        self.step(12)
+        asserts.assert_equal(reportedOnTimeValueList[2].value, 0, "Unexpected last OnTime report")
+
+        self.step(13)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OnTime]
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: Addressed

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 1 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: Initially I did not notice that it had been split, so once I noticed I added a note to the PR mentioning that I have not added this in OnOffLightingCluster yet. I'm getting a DUT setup which are running all-devices that I can use to verify.

Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: On the QuieterReporting.h, I'm aware that this is used in e.g. LevelControl and if we strongly think this is needed then I will try to use that instead, but it seemed like the simple path here was to specifically mark the scenarios where reporting is not needed directly, since it is fairly clear.

File: src/python_testing/TC_OO_2_8.py (hunk 2 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py:25]: I will get that added as well. Do we have any existing tests which are running against this already? I assume it will be "run 2: and then use ${ALL_DEVICES_APP}?

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 3 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:553]: Indeed, this is a loophole. It is based on the same rules as RemaningTime in LevelControl. The rationale here is that anything less than 10 is under a second, so the change is considered too small to bother reporting. @tennessee-google do you have any thoughts here?

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 4 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:654]: No, not every second. It is a countdown and these we only report the initial values or significant value changes and 0. It will be up to the client to keep track of the counting. The every second rule is typically for unpredictable values.

I have no strong opinion, I didn't want to introduce another if check when it was already there, but if we like the more compact code, I'm not going to fight it 😄 We just need to use kIfChanged instead of kYes, but else it should be fine.

File: src/python_testing/TC_OO_2_8.py (hunk 2 above)
Comment (andy31415) [reply src/python_testing/TC_OO_2_8.py:25]: https://github.com/project-chip/connectedhomeip/blob/14e806bf827a16a4c404c46ad3389250a52fb209/src/python_testing/TC_OCC_2_3.py#L34 seems to be an example.

Note that all-devices uses the Cluster.cpp (i.e. code driven) version.

File: src/python_testing/TC_OO_2_8.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+#       --storage-path admin_storage.json
+#       --commissioning-method on-network
+#       --discriminator 1234
+#       --passcode 20202021
+#       --PICS src/app/tests/suites/certification/ci-pics-values
+#       --endpoint 1
+#       --trace-to json:${TRACE_TEST_JSON}.json
+#       --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
+#     factory-reset: true
+#     quiet: true
```
Comment (ReneJosefsen) [inline src/python_testing/TC_OO_2_8.py:37]:
```suggestion
#     quiet: true
#   run2:
#     app: ${ALL_DEVICES_APP}
#     app-args: --device on-off-light --discriminator 1234 --KVS kvs1
#     script-args: >
#       --storage-path admin_storage.json
#       --commissioning-method on-network
#       --discriminator 1234
#       --passcode 20202021
#       --PICS src/app/tests/suites/certification/ci-pics-values
#       --trace-to json:${TRACE_TEST_JSON}.json
#       --trace-to perfetto:${TRACE_TEST_PERFETTO}.perfetto
#       --endpoint 1
#     factory-reset: true
#     quiet: true
```

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 9]
```diff
@@ -182,13 +182,27 @@ DataModel::ActionReturnStatus OnOffLightingCluster::WriteImpl(const DataModel::W
...
-        return persistence.DecodeAndStoreNativeEndianValue(request.path, decoder, mStartUpOnOff);
+        app::DataModel::ActionReturnStatus result =
+            persistence.DecodeAndStoreNativeEndianValue(request.path, decoder, mStartUpOnOff);
+        if (!result.IsSuccess())
+        {
+            return result;
+        }
+
+        NotifyAttributeChanged(request.path.mAttributeId);
+        return Status::Success;
```
Comment (andy31415) [inline src/app/clusters/on-off-server/OnOffLightingCluster.cpp]:
```cpp
        if (result.IsSuccess() && !result.IsNoOpSuccess())
        {
            NotifyAttributeChanged(request.path.mAttributeId);
        }

        return result;
```

We need to notify only if the change is not a noop (i.e. we do not notify if value is unchanged).

or maybe this is better:

```suggestion
        NotifyAttributeChangedIfSuccess(request.path.mAttributeId, result);
        return result;
```

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 10]
```diff
@@ -288,11 +316,13 @@ void OnOffLightingCluster::TimerFired()
 
         // TIMED_ON state: we decrement OnTime  to see if we need to turn off
         mOnTime--;
-        NotifyAttributeChanged(Attributes::OnTime::Id);
 
         // If timer is not yet 0, update the timer and keep going. Otherwise move to off state.
         VerifyOrReturn(mOnTime == 0, UpdateTimer());
 
+        // Only notify is OnTime reaches 0
+        NotifyAttributeChanged(Attributes::OnTime::Id);
```
Comment (andy31415) [inline src/app/clusters/on-off-server/OnOffLightingCluster.cpp:308]: Can we have unit tests on these?

Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.cpp:308]: I need to see if we have other instances of unit tests which validates notifications, but indeed if we can then we should. I intentionally started with just testing this against all-clusters and all-devices, I I will admit I did not focus on the unit tests (yet) 😄

Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.cpp:321]: Unit tests has been added

File: src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp [hunk 11]
```diff
@@ -531,6 +531,272 @@ TEST_F(TestOnOffLightingCluster, TestOffWaitTime)
     EXPECT_EQ(offWaitTime, 0);
 }
 
+TEST_F(TestOnOffLightingCluster, TestOnTimeAndOffWaitTimeReporting)
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp:534]: For the Code Duplication, this is IMO ok to have better readability of the test. The rest have been sorted

File: src/python_testing/TC_OO_2_8.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+        await self.send_single_cmd(Clusters.OnOff.Commands.Off())
+
+        self.step(21)
+        log.info("Test waits for 10 seconds")
+        await asyncio.sleep(10)
+
+        self.step(22)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OnTime]
+        reportedOnTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OnTime]
+        log.info(f'OnTime reports: {reportedOnTimeValueList}')
```
Comment (ReneJosefsen) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
```

File: src/python_testing/TC_OO_2_8.py [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        asserts.assert_less_equal(reportedOnTimeValueList[0].value, 100, "Unexpected first OnTime report")
+        asserts.assert_almost_equal(reportedOnTimeValueList[0].value, 100, delta=10, msg="Unexpected first OnTime report")
+
+        self.step(24)
+        asserts.assert_equal(reportedOnTimeValueList[1].value, 0, "Unexpected last OffWaitTime report")
+
+        self.step(25)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OffWaitTime]
+        reportedOffWaitTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OffWaitTime]
+        log.info(f'OffWaitTime reports: {reportedOffWaitTimeValueList}')
```
Comment (ReneJosefsen) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
```

File: src/python_testing/TC_OO_2_8.py [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+        asserts.assert_less_equal(reportedOnTimeValueList[1].value, 150, "Unexpected second OnTime report")
+        asserts.assert_almost_equal(reportedOnTimeValueList[1].value, 150, delta=10, msg="Unexpected second OnTime report")
+
+        self.step(12)
+        asserts.assert_equal(reportedOnTimeValueList[2].value, 0, "Unexpected last OnTime report")
+
+        self.step(13)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OffWaitTime]
+        reportedOffWaitTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OffWaitTime]
+        log.info(f'OffWaitTime reports: {reportedOffWaitTimeValueList}')
```
Comment (ReneJosefsen) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
```

File: src/python_testing/TC_OO_2_8.py [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        await self.send_single_cmd(cmd)
+
+        self.step(8)
+        log.info("Test waits for 20 seconds")
+        await asyncio.sleep(20)
+
+        self.step(9)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OnTime]
+        reportedOnTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OnTime]
+        log.info(f'OnTime reports: {reportedOnTimeValueList}')
```
Comment (ReneJosefsen) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
```

File: src/python_testing/TC_OO_2_8.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+        await asyncio.sleep(20)
+
+        self.step(9)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OnTime]
+        reportedOnTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OnTime]
+        asserts.assert_equal(count, 3, "Unexpected number of OnTime reports")
+
+        self.step(10)
+        asserts.assert_less_equal(reportedOnTimeValueList[0].value, 100, "Unexpected first OnTime report")
+        asserts.assert_almost_equal(reportedOnTimeValueList[0].value, 100, delta=10, msg="Unexpected first OnTime report")
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: This is due to the delta being +/- but in this case we only allow the value to be lower, so the delta needs to be 10. The two checks combined enforces that it is not allowed to be larger than the defined value.

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 17]
```diff
@@ -162,19 +162,19 @@ DataModel::ActionReturnStatus OnOffLightingCluster::ReadAttribute(const DataMode
...
     {
     case Attributes::OnTime::Id: {
         uint16_t value;
         ReturnErrorOnFailure(decoder.Decode(value));
         VerifyOrReturnValue(mOnTime != value, DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp);
+
+        if (abs(mOnTime - value) > mValueDeltaReportTrigger)
+        {
+            NotifyAttributeChanged(request.path.mAttributeId);
+        }
```
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.cpp]: This is intentional, a change smaller than 10 is considered insignificant and by skipping the reporting of the value, we also skip the reporting of reaching 0, since a client already thinks it is 0 based on its latest report.

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 3 above)
Comment (leorozendaal) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:553]: looking at the spec text "When it changes, with a delta larger than 10, **caused by a write or invoke of the
OnWithTimedOff command**", the case of multiple client actions which are <10 but collectively >10 is not covered by the spec. "Multiple updates in 1 s" is not in the spec... So this script PR matches the spec.

Implementing something which combines several <10 updates into a report would be not spec compliant 🤷 

For the avoidance of doubt: let's not try to make the spec and the script  overly complex for such corner case.

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 1 above)
Comment (cecille) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: I suspect QuieterReporting will simplify the implementation here and I THINK it also handles the weird scenario from below. Aka, I recommend using it, but I'm probably not going to die on this hill.

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 3 above)
Comment (cecille) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:553]: does the on-off-server.cpp suppress reports on writes of less than 10? I think these are different code paths, but I admit that I haven't been following the details of the code-driven / codegen split too closely.

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 18]
```diff
@@ -162,19 +162,19 @@ DataModel::ActionReturnStatus OnOffLightingCluster::ReadAttribute(const DataMode
...
-                                                              AttributeValueDecoder & decoder)
 {
     switch (request.path.mAttributeId)
     {
     case Attributes::OnTime::Id: {
         uint16_t value;
         ReturnErrorOnFailure(decoder.Decode(value));
         VerifyOrReturnValue(mOnTime != value, DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp);
+
+        if (abs(mOnTime - value) > mValueDeltaReportTrigger)
```
Comment (cecille) [inline src/app/clusters/on-off-server/OnOffLightingCluster.cpp]: Do these need to notify if the attribute is written to 0?

File: src/python_testing/TC_OO_2_8.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+
+    def pics_TC_OO_2_8(self):
+        """This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["OO.S", "OO.S.F00"]
+
+    def steps_TC_OO_2_8(self) -> list[TestStep]:
+        THcommand = "TH sends the command"
+        return [TestStep(1, test_plan_support.commission_if_required(), is_commissioning=True),
+                TestStep(2, f"{THcommand} Off command", test_plan_support.verify_success()),
+                TestStep(3, "Set up a subscription wildcard subscription for the On/Off Cluster, with MinIntervalFloor set to 0, MaxIntervalCeiling set to 30 and KeepSubscriptions set to false",
```
Comment (cecille) [inline src/python_testing/TC_OO_2_8.py]: If the ceiling is 30, could the device legally consolidate some of the reports?

File: src/python_testing/TC_OO_2_8.py [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+        await asyncio.sleep(20)
+
+        self.step(9)
+        count = sub_handler.attribute_report_counts[Clusters.OnOff.Attributes.OnTime]
+        reportedOnTimeValueList = sub_handler.attribute_reports[Clusters.OnOff.Attributes.OnTime]
+        asserts.assert_equal(count, 3, "Unexpected number of OnTime reports")
+
+        self.step(10)
+        asserts.assert_less_equal(reportedOnTimeValueList[0].value, 100, "Unexpected first OnTime report")
+        asserts.assert_almost_equal(reportedOnTimeValueList[0].value, 100, delta=10, msg="Unexpected first OnTime report")
```
Comment (cecille) [inline src/python_testing/TC_OO_2_8.py]: I think coderabbit is almost right - I think the delta might need to be 5. OTOH, given that you have two asserts anyway, you could also just swap this over to a greater than 95 check and not have to worry about it

File: src/python_testing/TC_OO_2_8.py [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+
+    def desc_TC_OO_2_8(self) -> str:
+        """Returns a description of this test"""
+        return "[TC-OO-2.8] Reporting requirements with (DUT as Server)"
+
+    def pics_TC_OO_2_8(self):
+        """This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["OO.S", "OO.S.F00"]
+
+    def steps_TC_OO_2_8(self) -> list[TestStep]:
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py]: any chance we can make this "ergonomic" to have an easier time reviewing that steps match text?

see https://github.com/project-chip/connectedhomeip/pull/71759

File: src/python_testing/TC_OO_2_8.py [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+        log.info("Test waits for 5 seconds")
+        await asyncio.sleep(5)
+
+        self.step(7)
+        cmd = Clusters.OnOff.Commands.OnWithTimedOff(onOffControl=0, onTime=150, offWaitTime=100)
+        await self.send_single_cmd(cmd)
+
+        self.step(8)
+        log.info("Test waits for 20 seconds")
+        await asyncio.sleep(20)
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py:106]: This is really bad for our CI performance.

Lets add it to https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/test_metadata.yaml#L242 as a "slow_tests"

File: src/python_testing/TC_OO_2_8.py (hunk 19 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: When min is 0? It properly could, I think we did discuss this even with min of 0, but it should not matter given the test is waiting for all the events in bulk and not one by one 🤔

File: src/python_testing/TC_OO_2_8.py (hunk 20 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: Doesn't it become 95 - 105 when delta is 10? But sure, I can do the asserts.assert_less_equal 100 and then an asserts.assert_greater_equal 95 and drop the delta. Should do the same thing.

File: src/app/clusters/on-off-server/OnOffLightingCluster.h [hunk 23]
```diff
@@ -125,13 +125,13 @@ class OnOffLightingCluster : public OnOffCluster, public TimerContext
     uint16_t mOnTime      = 0;
     uint16_t mOffWaitTime = 0;
 
+    uint8_t mValueDeltaReportTrigger = 10;
```
Comment (andy31415) [inline src/app/clusters/on-off-server/OnOffLightingCluster.h]: is there any reason for this to be an instance variable? Why not a constant `kValueDeltaReportTrigger` ?

File: src/python_testing/TC_OO_2_8.py [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+        self.step(2)
+        await self.send_single_cmd(Clusters.OnOff.Commands.Off())
+
+        # NOTE: added this sleep to let the DUT have some time to move
+        log.info("Test waits for 1 seconds")
+        await asyncio.sleep(1)
+
+        self.step(3)
+        sub_handler = AttributeSubscriptionHandler(expected_cluster=Clusters.OnOff)
+        await sub_handler.start(self.default_controller, self.dut_node_id, self.get_endpoint())
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py]: should we set parameters for subscriptions?


```suggestion
        await sub_handler.start(self.default_controller, self.dut_node_id, self.get_endpoint(), min_interval_sec=0, max_interval_sec=30, keepSubscriptions=False)
```

File: src/python_testing/TC_OO_2_8.py [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+        """This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["OO.S", "OO.S.F00"]
+
+    def steps_TC_OO_2_8(self) -> list[TestStep]:
+        THcommand = "TH sends the command"
+        return [TestStep(1, test_plan_support.commission_if_required(), is_commissioning=True),
+                TestStep(2, f"{THcommand} Off command", test_plan_support.verify_success()),
+                TestStep(3, "Set up a subscription wildcard subscription for the On/Off Cluster, with MinIntervalFloor set to 0, MaxIntervalCeiling set to 30 and KeepSubscriptions set to false",
+                         "Subscription successfully established"),
+                TestStep(4, f"	{THcommand} OnWithTimedOff command with AcceptOnlyWhenOn field set to 0, OnTime field set to 5 (0.5s) and OffWaitTime set to 5 (0.5s)",
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
                TestStep(4, f"{THcommand} OnWithTimedOff command with AcceptOnlyWhenOn field set to 0, OnTime field set to 5 (0.5s) and OffWaitTime set to 5 (0.5s)",
```

File: src/python_testing/TC_OO_2_8.py [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+                         test_plan_support.verify_success()),
+                TestStep(5, f"	{THcommand} OnWithTimedOff command with AcceptOnlyWhenOn field set to 0, OnTime field set to 100 (10s) and OffWaitTime set to 20 (2s)",
```
Comment (andy31415) [inline src/python_testing/TC_OO_2_8.py]:
```suggestion
                TestStep(5, f"{THcommand} OnWithTimedOff command with AcceptOnlyWhenOn field set to 0, OnTime field set to 100 (10s) and OffWaitTime set to 20 (2s)",
```

File: src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp [hunk 27]
```diff
@@ -523,6 +523,272 @@ TEST_F(TestOnOffLightingCluster, TestOffWaitTime)
...
+    // Timer should remain active to count down OnTime.
+    EXPECT_TRUE(mMockTimerDelegate.IsTimerActive(&mCluster));
+
+    // Advance clock
+    for (int i = 0; i < 11; ++i)
+    {
+        mMockTimerDelegate.AdvanceClock(System::Clock::Milliseconds32(100));
+    }
+
+    // Do not expect report since the initial change was not reported
```
Comment (andy31415) [inline src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp]: comment seems off: we have EXPECT_TRUE below

File: src/python_testing/TC_OO_2_8.py (hunk 24 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: It is the default values, but I have no problem in being explicit.

File: src/python_testing/TC_OO_2_8.py (hunk 25 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: This will be addressed in the transition to ergonomic

File: src/python_testing/TC_OO_2_8.py (hunk 26 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: This will be addressed in the transition to ergonomic

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 1 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: I'm not sure if it will be simpler, but it might indeed be required to address the write scenario on the "old" implementation. 😔

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp (hunk 18 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.cpp]: Yes, based on the spec text 0 is always reported so I added handling for this.

File: src/python_testing/TC_OO_2_8.py (hunk 21 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: Done

File: src/python_testing/TC_OO_2_8.py (hunk 22 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py:106]: Added

File: src/python_testing/TC_OO_2_8.py (hunk 24 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: Added

File: src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp (hunk 27 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/tests/TestOnOffLightingCluster.cpp]: The comment has been updated.

File: src/app/clusters/on-off-server/OnOffLightingCluster.h (hunk 23 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.h]: No, I changed it to a static const

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 3 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp:553]: After some additional testing it does not and I need to figure out how to sort that out. Looking at QuiterReporting I'm unsure if this can address this. Given this is on writes, which would likely not spam the network, as discussed, it might actually be worth changing the spec to state it shall report on any writes and only mandate the delta for commands.

File: src/app/clusters/on-off-server/codegen/on-off-server.cpp (hunk 1 above)
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/codegen/on-off-server.cpp]: Conclusion is that QuieterReporting is likely not going to solve issue of writes and we are better off tweaking the spec.

**Review by cecille — APPROVED** [review]
Talked to René offline and the questions about what happens on the write are actually fairly corner case and are distracting from the main purpose of this change, which is to reduce the countdown traffic. Whether a device does or does not report on write in a particular situation is not particularly relevant since it's only a single report and fully in control of the client. The spec could easily be relaxed for write and there would be functionally no change to the operational traffic in the real world.

Hence, approving. Let's land this, it's a good change, then deal with the corner cases of the write and the two different implementations after the test event, possibly by just relaxing the spec.

**Review by rbultman — APPROVED** [review]

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (ReneJosefsen) [reply src/app/clusters/on-off-server/OnOffLightingCluster.cpp]: The implementation has been modified to report all writes to the attribute, the delta check is only applied if done through a command. If another client invokes a command with a value tha that a delta less than 10, it is considered insignificant and other clients will get a report once it hits 0.

File: src/python_testing/TC_OO_2_8.py (hunk 20 above)
Comment (ReneJosefsen) [reply src/python_testing/TC_OO_2_8.py]: I have changed it to do a asserts.assert_less_equal and asserts.assert_greater_equal check instead of using the asserts.assert_almost_equal

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 28]
```diff
@@ -175,6 +175,7 @@ DataModel::ActionReturnStatus OnOffLightingCluster::WriteImpl(const DataModel::W
         uint16_t value;
         ReturnErrorOnFailure(decoder.Decode(value));
         VerifyOrReturnValue(mOnTime != value, DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp);
+
```
Comment (ReneJosefsen) [inline src/app/clusters/on-off-server/OnOffLightingCluster.cpp]:
```suggestion
```

File: src/app/clusters/on-off-server/OnOffLightingCluster.cpp [hunk 29]
```diff
@@ -183,6 +184,7 @@ DataModel::ActionReturnStatus OnOffLightingCluster::WriteImpl(const DataModel::W
         uint16_t value;
         ReturnErrorOnFailure(decoder.Decode(value));
         VerifyOrReturnValue(mOffWaitTime != value, DataModel::ActionReturnStatus::FixedStatus::kWriteSuccessNoOp);
+
```
Comment (ReneJosefsen) [inline src/app/clusters/on-off-server/OnOffLightingCluster.cpp]:
```suggestion
```
